### ISSUE TITLE
Python bindings (OB3): No module named 'openbabel._openbabel'

### ISSUE THREAD

**Issue by odioidentisti2** [openbabel/openbabel#2122]
Windows 10
Fresh install of Python 3.6 and OpenBabel 3.0.
Python bindings installed via 'pip install -U openbabel'

```
Microsoft Windows [Versione 10.0.17763.973]
(c) 2018 Microsoft Corporation. Tutti i diritti sono riservati.

C:\Users\alombardo>pip install -U openbabel
Requirement already up-to-date: openbabel in c:\python\python36\lib\site-packages (3.0.0)
You are using pip version 10.0.1, however version 20.0.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

C:\Users\alombardo>py
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 27 2018, 03:37:03) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from openbabel import openbabel as ob
Traceback (most recent call last):
  File "C:\Python\Python36\lib\site-packages\openbabel\openbabel.py", line 27, in swig_import_helper
    return importlib.import_module(mname)
  File "C:\Python\Python36\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'openbabel._openbabel'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python\Python36\lib\site-packages\openbabel\__init__.py", line 3, in <module>
    from . import openbabel
  File "C:\Python\Python36\lib\site-packages\openbabel\openbabel.py", line 30, in <module>
    _openbabel = swig_import_helper()
  File "C:\Python\Python36\lib\site-packages\openbabel\openbabel.py", line 29, in swig_import_helper
    return importlib.import_module('_openbabel')
  File "C:\Python\Python36\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ImportError: DLL load failed: Impossibile trovare il modulo specificato.
```

**Comment by baoilleach** [comment]
Hi. In the instructions above, you don't show the original installation by
pip so I can't tell if it has installed the wheel (the Windows binary), or
the source (which won't work). This could be a problem as you have an old
pip. I would recommend you update pip ("py -m pip -U pip" or something),
uninstall openbabel, and then reinstall and check that it is installing the
wheel.

**Comment by odioidentisti2** [comment]
It installed the wheel openbabel-3.0.0-cp36-cp36m-win_amd64.whl
I also tried to download it and install it manually from https://pypi.org/project/openbabel/#files but I got the same result

**Comment by baoilleach** [comment]
Does the file C:\Python\Python36\lib\site-packages\openbabel\_openbabel.pyd
exist?

If so, then the error may be related to difficulty finding the Open Babel
installation. Does "obabel" work at the command-line? Where is it
installed? Did you install the 64-bit version?

**Comment by odioidentisti2** [comment]
The file _openbabel.pyd doesn't exist: in the directory  C:\Python\Python36\lib\site-packages\openbabel\  I get only pybel.py, openbabel.py and __init__.py

The file obabel.exe is in C:\Program Files\OpenBabel-3.0.0, if I launch it from  inside its directory it works, from outside it doesn't.

I didn't get any error during the installation.
I don't know if it can be useful but openbabel has been installed by an administrator, the python bindings by a normal user.
Thank you

**Comment by baoilleach** [comment]
Actually _openbabel.pyd should be one-directory up. My bad. I think you'll
find it there.

The problem appears that your PATH doesn't include the folder containing
obabel.exe. The installer should have set this. In any case, you can just
set this at the command-line, e.g. set PATH=C:\Program
Files\OpenBabel-3.0.0;%PATH, prior to running Python. So try changing the
path, and then typing "obabel -:C -osmi" to see if it works. If this works,
try importing the Python library in the same command windows. If this
works, talk to your admin about adding it to the PATH permanently.

**Comment by odioidentisti2** [comment]
Yep, problem solved: it was just that entry missing in the PATH. :)

**Comment by pgg1610** [comment]
Hello, I am facing this issue with openbabel/3.1.1 installation which I doing from scratch following the recipe desribed on the website. 
python==3.8.5
openbabel==3.1.1

Error: 
```python
ModuleNotFoundError: No module named '_openbabel'
```

I can run open babel from command with user-defined PATH, PYTHONPATH, and Babel library paths. But having issues running it with python.

**Comment by fredrikw** [comment]
Most likely there is a path issue here as well. I would check in site-packages/openbabel and see if you have _openbabel.so in there. Also double check that you have the correct folder in sys.path.

**Comment by pgg1610** [comment]
Hello, @fredrikw thanks for the suggestion. I checked the path for my local install and I do see _openbabel.so in it. That said, I get the following error in python when I try to import openbabel: 

```python
from openbabel import openbabel as ob
ImportError: dynamic module does not define module export function (PyInit__openbabel)
```


```bash
ImportError                               Traceback (most recent call last)
Cell In [1], line 1
----> 1 from openbabel import openbabel as ob

File /users/repo/openbabel-install/lib/python3.8/site-packages/openbabel/__init__.py:3
      1 import sys
      2 import warnings
----> 3 from . import openbabel
      5 __version__ = "3.1.0"
      7 _thismodule = sys.modules[__name__]

File /users/repo/openbabel-install/lib/python3.8/site-packages/openbabel/openbabel.py:30
     28         except ImportError:
     29             return importlib.import_module('_openbabel')
---> 30     _openbabel = swig_import_helper()
     31     del swig_import_helper
     32 elif _swig_python_version_info >= (2, 6, 0):

File /users/repo/openbabel-install/lib/python3.8/site-packages/openbabel/openbabel.py:29, in swig_import_helper()
     27     return importlib.import_module(mname)
     28 except ImportError:
---> 29     return importlib.import_module('_openbabel')

File /apps/python/python-3.8.5/lib/python3.8/importlib/__init__.py:127, in import_module(name, package)
    125             break
    126         level += 1
--> 127 return _bootstrap._gcd_import(name[level:], package, level)
```

It seems that openbabel can find the local installation path `/lrlhps/users/repo/openbabel-install/lib/python3.8/site-packages/openbabel/` but then breaks down.

**Comment by fredrikw** [comment]
Ok, so
```(python)
import sys
print(sys.path)
```
contains `/users/repo/openbabel-install/lib/python3.8/site-packages/`?

Your `openbabel.py` looks different than mine (but is generated by swig so that may be ok), could you paste the first 40 lines of it so I can see?

**Comment by pgg1610** [comment]
> Ok, so
> 
> ```
> import sys
> print(sys.path)
> ```
> 
> contains `/users/repo/openbabel-install/lib/python3.8/site-packages/`?
> 
> Your `openbabel.py` looks different than mine (but is generated by swig so that may be ok), could you paste the first 40 lines of it so I can see?

Output from `sys.path`: 
['/data/dev/virtenv/auto3D/bin', 'users/repo/openbabel-install/build/lib', '/users/repo/openbabel-install/lib', '', '**users/repo/openbabel-install/lib/python3.8/site-packages**', '/data/c3tk/dev/virtenv/auto3D/lib/python3.8/site-packages', '/apps/python/python-3.8.5/lib/python38.zip', '/apps/python/python-3.8.5/lib/python3.8', '/apps/python/python-3.8.5/lib/python3.8/lib-dynload']

I do site-packages in the `sys.path` variable. 

Snippet of `openbabel.py` : 
```python
# This file was automatically generated by SWIG (http://www.swig.org).
# Version 3.0.12
#
# Do not make changes to this file unless you know what you are doing--modify
# the SWIG interface file instead.


import sys
if sys.platform.find("linux") != -1:
    dlflags = sys.getdlopenflags()
    import ctypes
    sys.setdlopenflags(dlflags | ctypes.RTLD_GLOBAL)



from sys import version_info as _swig_python_version_info
if _swig_python_version_info >= (3, 0, 0):
    new_instancemethod = lambda func, inst, cls: _openbabel.SWIG_PyInstanceMethod_New(func)
else:
    from new import instancemethod as new_instancemethod
if _swig_python_version_info >= (2, 7, 0):
    def swig_import_helper():
        import importlib
        pkg = __name__.rpartition('.')[0]
        mname = '.'.join((pkg, '_openbabel')).lstrip('.')
        try:
            return importlib.import_module(mname)
        except ImportError:
            return importlib.import_module('_openbabel')
    _openbabel = swig_import_helper()
    del swig_import_helper
elif _swig_python_version_info >= (2, 6, 0):
    def swig_import_helper():
        from os.path import dirname
        import imp
        fp = None
        try:
            fp, pathname, description = imp.find_module('_openbabel', [dirname(__file__)])
        except ImportError:
            import _openbabel
```

**Comment by pgg1610** [comment]
> Also, in your first comment, you mentioned the error `ModuleNotFoundError: No module named '_openbabel'` but in the last one it is a different error. Did you change anything in between?

Okay sorry for not providing context here.  I re-compiled openbabel using the following cmake and make commands: 
```bash
cmake ../openbabel-3.1.1 -DCMAKE_INSTALL_PREFIX=~/lrlhps_dir/repo/openbabel-install/  -DRUN_SWIG=ON -DEIGEN3_INCLUDE_DIR=/lrlhps/users/l017301/repo/openbabel-install/eigen-3.4.0 -DPYTHON_BINDINGS=ON

make && make install
```

**Comment by fredrikw** [comment]
I'm sorry but I don't follow all the way. There are a lot of different folders flying around. In the first install, you had prefix `~/repo/openbabel-install/`, in the second `~/lrlhps_dir/repo/openbabel-install/`. Did you remove everything from the first install before making the second?
Then you have `/users/repo/openbabel-install/lib/python3.8/site-packages` that seems to be a different path than both prefix folders? Also, in `sys.path` there seems to be missing a `/` before users at a few occasions?

**Comment by pgg1610** [comment]
> I'm sorry but I don't follow all the way. There are a lot of different folders flying around. In the first install, you had prefix `~/repo/openbabel-install/`, in the second `~/lrlhps_dir/repo/openbabel-install/`. Did you remove everything from the first install before making the second? Then you have `/users/repo/openbabel-install/lib/python3.8/site-packages` that seems to be a different path than both prefix folders? Also, in `sys.path` there seems to be missing a `/` before users at a few occasions?

oh oops my bad,  I have updated the command there. The directory structure is same as before I deleted the folders built during the installation and recompiled them.

**Comment by fredrikw** [comment]
But you still have a mixture of paths in the sys.path that you give above:
```(python)
[
'/data/dev/virtenv/auto3D/bin', 'users/repo/openbabel-install/build/lib', '/users/repo/openbabel-install/lib', '',
'users/repo/openbabel-install/lib/python3.8/site-packages', '/data/c3tk/dev/virtenv/auto3D/lib/python3.8/site-packages',
'/apps/python/python-3.8.5/lib/python38.zip', '/apps/python/python-3.8.5/lib/python3.8',
'/apps/python/python-3.8.5/lib/python3.8/lib-dynload'
]
````
And `~/repo/openbabel-install/´ does not seem to be in it. Have you changed your path manually? Is it really as you have pasted it? `'users/repo/openbabel-install'` and `'/users/repo/openbabel-install'` is the same path only if you start python from the root...

**Comment by pgg1610** [comment]
> 
These are exact paths. I was deleting folders for confidentiality. 

['/lrlhps/data/c3tk/dev/virtenv/auto3D/bin',
 '/lrlhps/users/<userid>/repo/openbabel-install/build/lib',
 '/lrlhps/users/<userid>/repo/openbabel-install/lib',
 '',
 '/lrlhps/users/<userid>/repo/openbabel-install/lib/python3.8/site-packages',
 '/lrlhps/data/c3tk/dev/virtenv/auto3D/lib/python3.8/site-packages',
 '/lrlhps/apps/python/python-3.8.5/lib/python38.zip',
 '/lrlhps/apps/python/python-3.8.5/lib/python3.8',
 '/lrlhps/apps/python/python-3.8.5/lib/python3.8/lib-dynload']

**Comment by pgg1610** [comment]
A follow up to this, I was able to compile a conda version of openbabel using `conda install -c conda-forge openbabel` and am trying on running a code using python. I got an error saying following: 

```python
ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /home/<user_id>/.conda/envs/auto3D/lib/python3.8/site-packages/openbabel/_openbabel.so)
```

which was solved by declaring the `LD_LIBRARY_PATH` variable pointing to the `env/lib` directory. I feel this seems to be a decent minimal working product to try out. I would revisit the path error which started my initial discussion at a later point.

Appreciate your help!
